### ISSUE TITLE
Proposal: Hyphenation support in WinUI

### ISSUE THREAD

**Issue by Marv51** [microsoft/microsoft-ui-xaml#4068]
# Proposal: Hyphenation support in WinUI
WPF had it, the web has it, WinUI needs it. Please add support for soft-hyphens to TextBlocks and other UI elements.
Currently words will be either wrapped to a new line at the letter that doesn't fit anymore, on a whitspace or not at all. 

A [soft-hyphen](https://en.wikipedia.org/wiki/Soft_hyphen) is the typographic correct way to indicate an appropriate place in a word to wrap to a new line. So if TextTrimming is set to WrapWholeWords, and if the word contains a unicode soft hypen  character (`U+00AD`, often encoded as `&shy;`), this character should be used to split the word on two lines with a hypen at the end of the first line. 

I believe good algorithms to determine whether to break at a soft-hypen are well understood and implemented in many projects (including Edge, WPF, Word, etc.).

## Rationale
Modern typography features are a major draw for developers to adopt WinUI, lets make sure WinUI is really up to par with other platforms.

This is an increasingly important feature if you take internationalization into account, where texts live independent of the box sizes they occupy in a resource file. Additionally, in certain languages (esp. German) long words are much more common and exacerbate the need for this feature.

## Stretch goal
A good follow up to this proposal would be a dictionary or algorithm based "auto-hyphenation" feature like some browsers and Word implement. This is especially important with justified text that is a bit longer (think a newspaper-like slim column text layout).

**Comment by omikhailov** [comment]
IMO, controls are not best place to implement advanced typography. If it will be ever implemented, it would be better to incapsulate such logic somewhere in the common place. For example, there could be some special RichString instead of string, or maybe even a set of specialized data types for HTML, markdown and other kinds of markup.

**Comment by Marv51** [comment]
@omikhailov Yes, of course. Supporting this would have to be implemented at a deep platform level, obviously I think so too. This is precisely the kind of change to the platform that WinUI 3 enables (or is supposed to enable).

This is a feature of how native-string-text is laid out on screen, using a standard more than three decades old control character. Exactly like the well supported non-breaking space character.

**Comment by mdtauk** [comment]
> @omikhailov Yes, of course. Supporting this would have to be implemented at a deep platform level, obviously I think so too. This is precisely the kind of change to the platform that WinUI 3 enables (or is supposed to enable).
> 
> This is a feature of how native-string-text is laid out on screen, using a standard more than three decades old control character. Exactly like the well supported non-breaking space character.
>
Xaml Typography already supports Text Wrapping and Text trimming with ellipses properties.  Hyphenation could be part of Text Wrapping or a new property added.

But would have to wait until WinUI 3's source is opened, or possibly the open source version of Direct Write is added into Reunion.

**Comment by yoshiask** [comment]
I've had this problem in one of my projects before, and using a zero-width space does work. WinUI will split the word at the ZWS if it needs to. It doesn't render a hyphen or anything, but if you just need a line break it's fine.

**Comment by omikhailov** [comment]
> Xaml Typography already supports Text Wrapping and Text trimming with ellipses properties. Hyphenation could be part of Text Wrapping or a new property added.
> 
> But would have to wait until WinUI 3's source is opened, or possibly the open source version of Direct Write is added into Reunion.

Ok, nevermind, I just understood it wrong. If it's about supporting _character_, this is really not so large feature and it can be part of current typography.

What I thought when I read the proposal is more common problem with rendering HTML markup with tags and _character entities_ like &nbsp; &shy; and other escape sequences encoding special symbols. Often, web applications store formatted text as HTML string in the database and when desktop app receives it from API, it is headache to render it and don't lose formatting.

**Comment by Marv51** [comment]
@omikhailov I have updated that sentence to be more clear about that.

**Comment by Marv51** [comment]
I opened this issue 3.5 years ago in the hope that this could be addressed, once WinUI 3 was published. 

We had 9 public releases of WinAppSDK since then, I still believe it is an important feature.

Maybe @codendone can help give this feature a chance to make it on the roadmap?
